src/connector/src/sink/remote.rs

```diff
@@ -0,0 +1,574 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use itertools::Itertools;
+use risingwave_common::array::{stream_chunk, StreamChunk};
+use risingwave_common::catalog::Schema;
+use risingwave_common::config::{MAX_CONNECTION_WINDOW_SIZE, STREAM_WINDOW_SIZE};
+use risingwave_common::types::{DataType, DatumRef, ScalarRefImpl};
+use risingwave_pb::connector_service::connector_service_client::ConnectorServiceClient;
+use risingwave_pb::connector_service::sink_config::table_schema::Column;
+use risingwave_pb::connector_service::sink_config::TableSchema;
+use risingwave_pb::connector_service::sink_task::write_batch::json_payload::RowOp;
+use risingwave_pb::connector_service::sink_task::write_batch::{JsonPayload, Payload};
+use risingwave_pb::connector_service::sink_task::{
+    Request as SinkRequest, StartEpoch, StartSink, SyncBatch, WriteBatch,
+};
+use risingwave_pb::connector_service::{SinkConfig, SinkResponse, SinkTask};
+use risingwave_pb::data;
+use risingwave_pb::data::data_type::TypeName;
+use serde_json::Value;
+use serde_json::Value::Number;
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use tokio_stream::StreamExt;
+use tonic::transport::{Channel, Endpoint};
+use tonic::{Request, Status, Streaming};
+
+use crate::sink::{Result, Sink, SinkError};
+
+pub const VALID_REMOTE_SINKS: [&str; 2] = ["jdbc", "file"];
+
+pub fn is_valid_remote_sink(sink_type: String) -> bool {
+    return VALID_REMOTE_SINKS.contains(&sink_type.as_str());
+}
+
+#[derive(Clone, Debug)]
+pub struct RemoteConfig {
+    pub sink_type: String,
+    pub properties: HashMap<String, String>,
+}
+
+impl RemoteConfig {
+    pub fn from_hashmap(values: HashMap<String, String>) -> Result<Self> {
+        let sink_type = values
+            .get("connector")
+            .expect("sink type must be specified")
+            .to_string();
+
+        if !is_valid_remote_sink(sink_type.clone()) {
+            return Err(SinkError::Config(format!("invalid sink type: {sink_type}")));
+        }
+
+        Ok(RemoteConfig {
+            sink_type,
+            properties: values,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum RemoteSinkResponseImpl {
```

Comment:
Since it is only used inside this file, we don't need to `pub` it. If it is used only in this file, we can remove the `RemoteSink` prefix to avoid making it too verbose. Besides, `ResponseStreamImpl` is better since this is a wrapper for stream.